clients/src/main/java/org/apache/kafka/common/requests/AlterConfigsRequest.java

```diff
@@ -104,98 +64,76 @@ public String value() {
+            resource -> new ConfigResource(
+                    ConfigResource.Type.forId(resource.resourceType()),
+                    resource.resourceName()),
+            resource -> new Config(resource.configs().stream()
+                    .map(entry -> new ConfigEntry(entry.name(), entry.value()))
+                    .collect(Collectors.toList()))));
     }
 
     public boolean validateOnly() {
-        return validateOnly;
+        return data.validateOnly();
     }
 
     @Override
     protected Struct toStruct() {
-        Struct struct = new Struct(ApiKeys.ALTER_CONFIGS.requestSchema(version()));
-        struct.set(VALIDATE_ONLY_KEY_NAME, validateOnly);
-        List<Struct> resourceStructs = new ArrayList<>(configs.size());
-        for (Map.Entry<ConfigResource, Config> entry : configs.entrySet()) {
-            Struct resourceStruct = struct.instance(RESOURCES_KEY_NAME);
-
-            ConfigResource resource = entry.getKey();
-            resourceStruct.set(RESOURCE_TYPE_KEY_NAME, resource.type().id());
-            resourceStruct.set(RESOURCE_NAME_KEY_NAME, resource.name());
-
-            Config config = entry.getValue();
-            List<Struct> configEntryStructs = new ArrayList<>(config.entries.size());
-            for (ConfigEntry configEntry : config.entries) {
-                Struct configEntriesStruct = resourceStruct.instance(CONFIG_ENTRIES_KEY_NAME);
-                configEntriesStruct.set(CONFIG_NAME, configEntry.name);
-                configEntriesStruct.set(CONFIG_VALUE, configEntry.value);
-                configEntryStructs.add(configEntriesStruct);
-            }
-            resourceStruct.set(CONFIG_ENTRIES_KEY_NAME, configEntryStructs.toArray(new Struct[0]));
-
-            resourceStructs.add(resourceStruct);
-        }
-        struct.set(RESOURCES_KEY_NAME, resourceStructs.toArray(new Struct[0]));
-        return struct;
+        return data.toStruct(version());
     }
 
     @Override
     public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable e) {
         ApiError error = ApiError.fromThrowable(e);
-        Map<ConfigResource, ApiError> errors = new HashMap<>(configs.size());
-        for (ConfigResource resource : configs.keySet())
-            errors.put(resource, error);
-        return new AlterConfigsResponse(throttleTimeMs, errors);
+        AlterConfigsResponseData data = new AlterConfigsResponseData()
+                .setThrottleTimeMs(throttleTimeMs);
+        for (AlterConfigsRequestData.AlterConfigsResource resource : this.data.resources()) {
+            data.responses().add(new AlterConfigsResponseData.AlterConfigsResourceResponse()
+                    .setResourceType(resource.resourceType())
+                    .setResourceName(resource.resourceName())
+                    .setErrorMessage(error.message())
+                    .setErrorCode(error.error().code()));
+        }
+        return new AlterConfigsResponse(data);
```

Comment:
This data construction seems better to be put in the `AlterConfigsResponse` constructor.

Reply:
Thanks, @abbccdda. There was such a constructor in the `AlterConfigsResponse` originally, but it used a `Map` parameter (and it seemed a bit pointless to me to create a map here just to iterate over it in the constructor). Are you suggesting to revert to that constructor, or have a signiture like `(List<AlterConfigsRequestData.AlterConfigsResource> resources, ApiError error)`?

Reply:
Yea, my suggestion would be to reuse the existing constructor as the construction of the `AlterConfigsResponseData` seems non trivial for a caller to do, compared with passing a map of errors.

Reply:
But it only has this one call site, so the benefit is not clear to me.